collect_app/src/test/java/org/odk/collect/android/preferences/JsonPreferencesGeneratorTest.kt

```diff
@@ -142,11 +153,11 @@ class JsonPreferencesGeneratorTest {
         settingsProvider.getAdminSettings("2").saveAll(adminPrefsForProjectX)
 
         // Verify the demo project
-        verifyJsonContent(jsonPreferencesGenerator.getJSONFromPreferences(), generalPrefsForDemoProject, adminPrefsForDemoProject)
+        verifyJsonContent(jsonPreferencesGenerator.getJSONFromPreferences(), generalPrefsForDemoProject, adminPrefsForDemoProject, projectDetails)
```

Comment:
This all really confused me as the test is using a real `SettingsProvider` which means it uses the real `CurrentProjectProvider` but then the actual object under test uses a mock `CurrentProjectProvider`. I think we need to clean that up as the boundaries of this test aren't quite defined.

One idea is that we create a test settings provider implementation (and pull out an interface to sit above the real and test one) that just uses `InMemSettings` under the hood and doesn't use the `CurrentProjectProvider` (it just returns different settings objects for different project Ids and a default one for no project Id). This would be really useful in a lot of places as we could use it instead of the real `SettingsProvider` and potentially avoid Robolectric as well (as we need it for `SharedPreferences`).

Reply:
> which means it uses the real CurrentProjectProvider but then the actual object under test uses a mock CurrentProjectProvider

why? I think there is no real `CurrentProjectProvider` just a mock.

I thought this test had a broader responsibility (testing generating jsons with real settings). It's not a test for one class (SettingsImporter for example) which then (I would agree) should rather have mocked settings.

Reply:
> why? I think there is no real CurrentProjectProvider just a mock.

The test interacts with a real `CurrentProjectProvider` through Dagger:

```
// Verify the 'Project X' project
        DaggerUtils.getComponent(ApplicationProvider.getApplicationContext<Collect>()).currentProjectProvider().setCurrentProject("2")
```

The `SettingsProvider` is also interacting with this real `CurrentProjectProvider` as well, whereas the `AppConfigurationGenerator` is interacting with a mock one. I think it's confusing that we need to provide setup for the `SettingsProvider` here which I think we trust (and have tested) to return the right settings objects. To me, it feels like a fake/stub `SettingsProvider` could be used here instead and that would make it clearer what the `AppConfigurationGenerator` is responsible for. The stub could return `SharedPreferencesSettings` if that feels like a more thorough test, but I'm not sure we'd need that. Does that make sense?

Reply:
Hmm yes I think I confused this test (which is about exporting prefs to json) with the one which decodes json and my thinking was wrong.
Ok I think mocking `SettingsProvider` makes sense:
https://github.com/getodk/collect/pull/4607/commits/d5dbd8160b995400c799d0f16a669fecf9d127b5
Do you like this version?

Reply:
Ah right! Yeah this looks great.